facades/entrypoints.md

```diff
@@ -0,0 +1,185 @@
+
+**(b)** The facade should promote the most commonly-used functions of the network provider.
+
+**(c)** The facade should fetch network parameters (network config) needed for construction transactions (e.g. chain ID) in a lazy manner. For example, it may fetch them when the first transaction is created. This works fine for JavaScript, as well, since transaction-creation functions are async anyway.
+
+```
+class NetworkEntrypoint:
+    // The constructor is not captured by the specs; it's up to the implementing library to define it.
+    // For example, it can be parametrized with: a network provider URL and kind (proxy, API)
+    constructor(url: str, kind: Optional[str]);
+
+    verify_transaction_signature(transaction: Transaction): bool;
+    verify_message_signature(message: Message): bool;
+
+    // Fetches the account nonce from the network.
+    recall_account_nonce(address: IAddress): uint64;
+
+    // Function of the network provider, promoted to the facade.
+    send_transaction(transaction: Transaction): string;
+
+    // Function of the network provider, promoted to the facade.
+    send_transactions(transaction: Transaction); Tuple[int, List[string]];
+
+    // Generic function to await a transaction on the network.
+    await_completed_transaction(transaction_hash: string): TransactionOnNetwork;
+
+    // Access to the underlying network provider.
+    get_network_provider(): INetworkProvider;
+
+    // Access to the individual controllers.
+    create_smart_contract_controller(abi: Optional[Abi]): SmartContractController;
+    get_transfers_controller(): TransfersController;
+    get_token_management_controller(): TokenManagementController;
+    get_delegation_controller(): DelegationController;
+    get_relayed_controller(): RelayedController;
+    get_account_management__controller(): AccountManagementController;
+```
+
+## Examples
+
+### Transfer native amounts and / or custom tokens
+
+```
+entrypoint = MainnetEntrypoint();
+controller = entrypoint.get_transfer_transactions_controller()
+
+sender = entrypoint.load_account_from_pem("alice.pem");
+sender.nonce = entrypoint.recall_account_nonce(sender.address);
+
+transaction = controller.create_transaction_for_transfer({
+    sender: sender,
+    nonce: sender.get_nonce_then_increment(),
+    receiver: Address.from_bech32("erd1..."),
+    native_amount: 1000000000000000000,
+    token_transfers: [];
+});
+
+// Broadcast transaction (option 1):
+transaction_hash = entrypoint.send_transaction(transaction);
+// Broadcast transaction (option 2):
+transaction_hash = entrypoint.get_network_provider().send_transaction(transaction);
+
+// Await transaction completion (option 1, await and specialized outcome parsing):
+parsed_outcome = controller.await_completed_transfer(transaction_hash);
+// Await transaction completion (option 2, simple await):
+transaction_on_network = entrypoint.await_completed_transaction(transaction_hash);
```

Comment:
Even if creation (and signing) of transactions is done by controllers, broadcasting isn't (see examples). 

All good? Alternative: have `send_transaction()` in each controller, instead of having it in the entrypoint?

Reply:
I think it's ok as is - maybe if we add them into controller would open discussions about consistency, eg: "why is send_transaction inside the controller but other network related suff like recall_account_nonce is in the entrypoint?"

Reply:
1. Maybe I'm too nitpicky and I keep coming back to old talks that maybe we already concluded, but like in this example, we get the `get_transfers_controller` only to use it once to create a transaction, then never use it again (cause everywhere else we're using the entrypoint). 
I will most probably just do `entrypoint.get_transfers_controller().create_transaction_for_transfer()` to shorten it for single use only and it wouldn't be the end of the world, but the idea of creating that intermediary object to get to the transaction is nagging me ever so slightly, and this leads me to the second point:

2. Is that a `controller`? Since it still looks more like a factory, just different in the way that I take it directly from the entrypoint rather than importing it myself. 
Much easier to get and use than the base sdk intricacies, nevertheless.

3. One thing as food for thought that I keep wondering for a while and asked before but maybe lost its point is whether we need to have separate entities for Address and Account. I would rather use Account everywhere with the particularity that some can have signing abilities (if private key is provided to them), others don't (if it only has a public key available like: Account.new_from_bech32("blabla")).
The reason I'm picking at this is because at chain level they're the same entity, but they're not interchangeable in code and they introduce a debateable unnecessary friction in the fact that I have to import them separately and use them differently based on where they go (if I'm a sender, I need to use the account; if i'm the receiver, I have to use the address). Not much effort involved into this, I know, but still an effort that has to be explained and understood.

4. Should we stick to `entrypoint` name?

Reply:
1. Fair point. In this last version we've thought of not having functions that create transactions directly on the entrypoint, but in the child controllers, instead. Debatable, though :+1: 

2. Indeed, it's a controller. The more lower-level factory does not returned signed transactions (the controller signs the transactions). The controller is also able to do a "specialized" await for a transaction's completion - await & properly parse its outcome (the lower level fashion is, now, to await for transaction completion using a transaction awaiter, then use a transaction outcome parser to recover & interpret returned data - e.g. logs).

3. I'd like to keep `Address` a different concept, as it is now. Especially because I'm thinking of https://github.com/multiversx/mx-sdk-specs/issues/36.

4. Somehow better than `facade`.

Reply:
2. Understood the point and wasn't aware of those things without seeing the background. For example, `entrypoint.await_completed_transaction` leave the impression that the await is done here, while the signing is invisible, so the only thing the dev _sees_ is the factory/builder element :) 
Maybe we complicate the names too much with the patterns, don't know.

Reply:
Noted :+1:

Reply:
oh God. I completely missed this. I think this is reopening up @ccorcoveanu's comment above for debate on consistency.